optimizer/lib/transformers/BlurImagePlaceholders.js

```diff
@@ -0,0 +1,160 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const sizeOf = require('image-size');
+const jimp = require('jimp');
+
+/**
+ * Adds placeholders for all AMP images and posters that are blurry versions of
+ * the corresponding original source. The blur will be displayed as the
+ * <amp-img> is rendering, and will fade out once the element is loaded.
+ */
+class BlurImagePlaceholders {
+  /**
+   * Parses the document to add blurred placedholders in all appropriate
+   * locations.
+   * @param tree A parse5 treeAdapter.
+   * @return {Array} An array of promises that all represents the resolution of
+   * a blurred placeholder being added in an appropriate place.
+   */
+  transform(tree) {
+    const html = tree.root.firstChildByTag('html');
+    const body = html.firstChildByTag('body');
+    const promises = [];
+    for (let node = body; node !== null; node = node.nextNode()) {
+      const {tagName} = node;
+      let src;
+      if (tagName === 'amp-img') {
+        src = node.attribs.src;
+      }
+      if (tagName === 'amp-video' && node.attribs.poster) {
+        src = node.attribs.poster;
+      }
+      if (src && !this.hasPlaceholder_(node)) {
+        promises.push(this.addBlurryPlaceholder_(tree, src).then(imgChild => {
+          node.appendChild(imgChild);
+        }));
+      }
+    }
+    return Promise.all(promises);
+  }
+
+  /**
+   * Adds a child image that is a blurry placeholder.
+   * @param tree A parse5 treeAdapter.
+   * @param {String} src The image that the bitmap is based on.
+   * @return {!Promise} A promise that signifies that the img has been updated
+   * to have correct attributes to be a blurred placeholder.
+   * @private
+   */
+  addBlurryPlaceholder_(tree, src) {
+    const img = tree.createElement('img');
+    img.attribs.src = src;
+    img.attribs.class = 'i-amphtml-blur';
+    img.attribs.placeholder = '';
+    return this.getDataURI_(img).then(() => {
+      return img;
+    }).catch(err => {
+      console.error("Transformer error during the calculation of bitmap " +
```

Comment:
Minor: Would be nice to prefix this error with the transformer name for disambiguation.